interpreter/valid/valid.ml

```diff
@@ -157,17 +157,63 @@ let check_comptype (c : context) (ct : comptype) at =
     check_resulttype c ts1 at;
     check_resulttype c ts2 at
 
-(* TODO: Make sure the describes and descriptor clauses match. *)
 let check_desctype (c : context) (dt : desctype) at =
-  match dt with
-  | DescT (ht1, ht2, ct) -> check_comptype c ct at
-
-let check_desctype_sub (c : context) (dt : desctype) (dti : desctype) x xi at =
-  (* TODO: check rules *)
+  (match dt with
+  | DescT (None, None, _) -> ()
+  | DescT (_, _, StructT _) -> ()
+  | DescT (_, Some _, _) ->
+    error at "descriptor type must be a struct"
+  | DescT (Some _, _, _) ->
+    error at "described type must be a struct");
   let DescT (_, _, ct) = dt in
-  let DescT (_, _, cti) = dti in
-  require (match_comptype c.types ct cti) at ("sub type " ^ I32.to_string_u x ^
-      " does not match super type " ^ I32.to_string_u xi)
+  check_comptype c ct at
+
+let check_desctype_sub (c : context) (dt : desctype) (dt' : desctype) x x' at =
+  let DefT (rt, _) = type_ c (x @@ at) in
+  let DefT (rt', _) = type_ c (x' @@ at) in
+  let DescT (ut1, ut2, ct) = dt in
+  let DescT (ut1', ut2', ct') = dt' in
+  match ut1, ut1' with
+  | (Some ut1, Some ut1') ->
+    require (match_typeuse c.types ut1 ut1' rt rt') at ("described type " ^
+        string_of_typeuse ut1 ^ " does not match " ^ string_of_typeuse ut1')
+  | (Some _, None) | (None, Some _) ->
+    error at ("sub type " ^ I32.to_string_u x ^ " does not match super type " ^
+        I32.to_string_u x')
+  | (None, None) -> ();
+  match ut2, ut2' with
```

Comment:
This doesn't mean what you may think it does. You probably want to put some parens around the prior match, otherwise this match is just part of the last branch.

So the conditions on `ut2`/`ut2'` are actually not checked in the presence of descriptor clauses, nor the ones on `ct`/`ct'`. That may also suggest that we lack sufficient negative tests.